Homework 9

Dragos-Andrei-75 opened the pull request:
Defined endpoints:
- searchBooks
    - searches for books based on title and/or author and returns all books that match;
- reserveBook
    - finds a book by ID, searches through its list of exemplars and returns the first available exemplary;

Dragos-Andrei-75 committed:
Homework 9

Defined endpoints:
- searchBooks
    - searches for books based on title
      and/or author and returns all
      books that match;
- reserveBook
    - finds a book by ID, searches
      through its list of exemplars and
      returns the first available
      exemplary;

marcopacurariu3 commented on src/main/java/com/example/SpringBookstore/Status.java:
```diff
@@ -0,0 +1,8 @@
+package com.example.SpringBookstore;
+
+public enum Status {
```
hai sa fie intr-un pachet reservation sau punem "ReservationStatus", altfel nu stim despre ce status e vorba doar citind acest enum

marcopacurariu3 commented on src/main/java/com/example/SpringBookstore/controller/ReservationController.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    @GetMapping
+    public ResponseEntity<?> searchBooks(@RequestParam(required = false) String title, @RequestParam(required = false) String author, @RequestParam Integer pageNumber, @RequestParam Integer numberOfElements) {
```
hai sa ii spunem "pageSize" lui "numberOfElements"

marcopacurariu3 commented on src/main/java/com/example/SpringBookstore/controller/ReservationController.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    @PostMapping(path = "/{userID}/{bookID}")
+    public ResponseEntity<?> reserveBook(@PathVariable(name = "userID") Long userID, @PathVariable(name = "bookID") Long bookID, @RequestParam LocalDate startDate, @RequestParam LocalDate endDate) {
```
statDate & endDate as zice sa le punem mandatory, si atunci le-as cam pune intr-un body, nu query parameters

marcopacurariu3 commented on src/main/java/com/example/SpringBookstore/entities/Book.java:
```diff
@@ -123,4 +123,18 @@ public Library getLibrary() {
         this.library = library;
     }
+
+    public void addExemplary(Exemplary exemplary) {
```
super :)

marcopacurariu3 commented on src/main/java/com/example/SpringBookstore/entities/Exemplary.java:
```diff
@@ -20,11 +20,14 @@ public class Exemplary {
         return id;
     }
 
-    public void setId(Long id) {
```
e ok cu "d" mic, daca campul e si el cu "d" mic

marcopacurariu3 commented on src/main/java/com/example/SpringBookstore/entities/Librarian.java:
```diff
@@ -34,11 +34,11 @@ public class Librarian {
     private Library library;
 
-    public Long getId() {
+    public Long getID() {
```
acum vad ca le-ai schimbat pe toate. ce sa zic, nu te-as mai pune acum sa le schimbi inapoi, lasa... :)

marcopacurariu3 commented on src/main/java/com/example/SpringBookstore/entities/Reservation.java:
```diff
@@ -0,0 +1,81 @@
+    @Column(name = "STATUS")
+    private Status status;
+
+    @OneToOne(fetch = FetchType.LAZY)
```
e @ManyToOne aici. un exemplar va avea mai multe rezervari, nu o singura rezervare

marcopacurariu3 commented on src/main/java/com/example/SpringBookstore/mapper/LibrarianMapper.java:
```diff
@@ -7,7 +7,7 @@ public class LibrarianMapper {
         Librarian librarian = new Librarian();
 
-        librarian.setId(librarianDTO.getId());
+        librarian.setID(librarianDTO.getID());
```
sa lasi asa, ca si isbn le-ai facut cu mari, e ok

marcopacurariu3 replied:
ca macar e consistent, all good

marcopacurariu3 commented on src/main/java/com/example/SpringBookstore/mapper/ReservationMapper.java:
```diff
@@ -0,0 +1,22 @@
+        reservationDTO.setStatus(reservation.getStatus());
+        reservationDTO.setExemplaryDTO(ExemplaryMapper.exemplary2ExemplaryDTO(reservation.getExemplary()));
+
+        if (reservation.getUser() != null) {
```
se poate intampla asta? intreb ca fata de exemplary nu ai acea verificare, deci intreb ce te-ai facut cu gandul ca user-ul poate sa lipseasca, dar exemplary nu. ca mie mi se pare ca ambele ar trebui sa fie prezente

marcopacurariu3 commented on src/main/java/com/example/SpringBookstore/repositories/BookRepository.java:
```diff
@@ -1,9 +1,20 @@
 
 @Repository
 public interface BookRepository extends JpaRepository<Book, Long> {
+    @Query(value = """
```
cool :)

marcopacurariu3 commented on src/main/java/com/example/SpringBookstore/repositories/ExemplaryRepository.java:
```diff
@@ -1,9 +1,46 @@
+    */
+
+    //Native Query - it causes more trouble than it's worth in case you also want to specify a cascade type for the Exemplary field inside the Reservation class;
+    @Query(value = """
```
nu trebuia sa tii cont si de perioade in acest query?

ps: ar trebui sa poti pune limit si in query-ul de Hibernate, nu doar in cel nativ de postgresql. nu a mers?

marcopacurariu3 commented on src/main/java/com/example/SpringBookstore/repositories/ExemplaryRepository.java:
```diff
@@ -1,9 +1,46 @@
+            LEFT JOIN reservations aReservation ON aExemplary.id = aReservation.exemplary_id
+            WHERE aExemplary.book_id = :bookID
+            AND
+            (aReservation.id IS NULL OR aReservation.status = 'FINISHED')
```
not sure daca e corect. nu trebuia aici un query de genul: selecteaza-mi toate exemplarele (ma rog, limit 1) pentru care nu exista exemplare ce au vreo rezervare?

marcopacurariu3 commented on src/main/java/com/example/SpringBookstore/service/LibrarianService.java:
```diff
@@ -65,8 +65,9 @@ public List<Librarian> findAll() {
     public void delete(Long librarianID) {
-        Librarian librarian = findByID(librarianID);
-        librarianRepository.delete(librarian);
+        if (librarianRepository.existsById(librarianID)) {
```
stiu ca in alte parti aruncai si exceptie daca nu exista => consistency

marcopacurariu3 commented on src/main/java/com/example/SpringBookstore/service/ReservationService.java:
```diff
@@ -0,0 +1,74 @@
+        bookRepository.findById(bookID)
+                .orElseThrow(() -> new EntityNotFoundException("Book with ID " + bookID + " not found."));
+
+        Exemplary exemplary = exemplaryRepository.reserveExemplary(bookID)
```
looks good, dar, cum ziceam, trebuie sa cauti un exemplare ne-book-uit in acea perioada, nu ai tinut cont de perioade

Dragos-Andrei-75 committed:
Addressed All Issues from the Previous Review

Addressed all issues from the previous review made by Marco Păcurariu.

Added the Bookstore.postman_collection.json file into the resources folder.

Bookstore.postman_collection.json contains request for testing each endpoint that exist within the project.
